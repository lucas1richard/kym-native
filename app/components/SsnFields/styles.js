import { StyleSheet } from 'react-native';
import theme from 'theme/iftheme';

const styles = StyleSheet.create({
  wrapper: {
    marginBottom: 40,
    backgroundColor: '#fff',
    width: '100%'
  },
  flexWrapper: {
    flexDirection: 'row',
    justifyContent: 'space-between'
  },
  input: {
    backgroundColor: '#fff',
    position: 'relative',
    fontWeight: '300',
    height: 24
    // borderColor: theme.darkGray,
    // borderWidth: 1,
    // padding: 8,
    // borderRadius: 4
  },
  textInputWrapper: {
    borderRadius: 4,
    padding: 8,
    borderColor: theme.darkGray,
    borderWidth: 1
  },
  textInputWrapperError: {
    borderRadius: 4,
    padding: 8,
    borderColor: 'red',
    borderWidth: 1
  }
});

export default styles;
