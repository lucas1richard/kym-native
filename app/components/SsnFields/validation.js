import { ssn } from 'utilities/validation';

const ssnValidation = (value) => {
  if (!value) {
    return 'Required';
  }
  return ssn(value);
};

export default ssnValidation;
