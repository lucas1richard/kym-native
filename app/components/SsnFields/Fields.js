/**
* Fields
*/
import React from 'react';
import { View, Text, TextInput } from 'react-native';
import PropTypes from 'prop-types';
import theme from 'theme/iftheme';
import styles from './styles';

export class Fields extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  constructor() {
    super();
    this.state = {
      ssn1: '',
      ssn2: '',
      ssn3: ''
    };
  }

  change(value, key, max, nextComp) {
    Promise.resolve(this.setState({ [key]: value })).then(() => {
      this.props.input.onBlur();
      this.props.input.onChange(`${this.state.ssn1}${this.state.ssn2}${this.state.ssn3}`);
      if (value.length === max && nextComp) {
        nextComp.focus();
      }
    });
  }

  render() {
    const { meta } = this.props;
    const error = meta.touched && meta.error;
    const valid = meta.touched && !meta.error;
    let inputWrapper = !error ? styles.textInputWrapper : styles.textInputWrapperError;
    if (valid) {
      inputWrapper = [styles.textInputWrapper, { borderColor: theme.green }];
    }
    return (
      <View style={styles.wrapper}>
        <Text valid={valid}>Social Security Number</Text>
        <View style={styles.flexWrapper}>
          <View style={[inputWrapper, { width: '33%' }]}>
            <TextInput
              onChangeText={(value) => this.change(value, 'ssn1', 3, this.ssn2)}
              value={this.state.ssn1}
              maxLength={3}
              error={error}
              style={styles.input}
              keyboardType="numeric"
              placeholder="###"
              valid={valid}
              returnKeyType="done"
            />
          </View>
          <View style={[inputWrapper, { width: '25%' }]}>
            <TextInput
              onChangeText={(value) => this.change(value, 'ssn2', 2, this.ssn3)}
              value={this.state.ssn2}
              maxLength={2}
              ref={(comp) => { this.ssn2 = comp; }}
              error={error}
              style={styles.input}
              keyboardType="numeric"
              placeholder="##"
              valid={valid}
              returnKeyType="done"
            />
          </View>
          <View style={[inputWrapper, { width: '37%' }]}>
            <TextInput
              onChangeText={(value) => this.change(value, 'ssn3', 4)}
              value={this.state.ssn3}
              maxLength={4}
              ref={(comp) => { this.ssn3 = comp; }}
              error={error}
              style={styles.input}
              keyboardType="numeric"
              placeholder="####"
              valid={valid}
              returnKeyType="done"
            />
          </View>
        </View>
        {error && <Text show={error} style={{ color: 'red' }}>Not a valid SSN</Text>}
      </View>
    );
  }
}

Fields.propTypes = {
  input: PropTypes.object,
  meta: PropTypes.object
};

export default Fields;
