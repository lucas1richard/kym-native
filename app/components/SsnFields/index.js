/**
* SsnFields
*/
import React from 'react';
import { Field } from 'redux-form/immutable';

import Fields from './Fields';
import validateSsn from './validation';

export class SsnFields extends React.PureComponent { // eslint-disable-line react/prefer-stateless-function
  render() {
    return (
      <Field
        name="ssn"
        component={Fields}
        validate={validateSsn}
      />
    );
  }
}

export default SsnFields;
