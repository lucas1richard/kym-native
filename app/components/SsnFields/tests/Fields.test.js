import React from 'react';
import { shallow, mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import SsnFields, { FirstInput, SecondInput, ThirdInput } from '../Fields';

describe('<Fields />', () => {
  const renderedComponent = (props = {}) => shallow(<SsnFields {...props} />);

  const props = {
    meta: {
      touched: false,
      error: 'Error'
    },
    input: {
      onBlur: jest.fn(),
      onChange: jest.fn()
    }
  };
  it('should render without blowing up', () => {
    expect(toJson(renderedComponent(props))).toMatchSnapshot();
    props.meta.touched = true;
    expect(toJson(renderedComponent(props))).toMatchSnapshot();
  });

  it('should handle the change class function', () => {
    const component = mount(<SsnFields {...props} />);
    const firstInput = component.find(FirstInput);
    const secondInput = component.find(SecondInput);
    const thirdInput = component.find(ThirdInput);

    firstInput.simulate('change', { target: { value: '11' } });
    expect(component.state('ssn1')).toEqual('11');

    secondInput.simulate('change', { target: { value: '12' } });
    expect(component.state('ssn2')).toEqual('12');

    thirdInput.simulate('change', { target: { value: '123' } });
    expect(component.state('ssn3')).toEqual('123');
  });
});
