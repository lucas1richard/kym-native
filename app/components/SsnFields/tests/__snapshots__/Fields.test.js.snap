// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Fields /> should render without blowing up 1`] = `
<styled.div>
  <Styled(styled.div)
    error={false}
    valid={false}
  >
    SSN
  </Styled(styled.div)>
  <styled.div>
    <Styled(styled.input)
      error={false}
      maxLength={3}
      onChange={[Function]}
      valid={false}
      value=""
    />
    <Styled(styled.input)
      error={false}
      innerRef={[Function]}
      maxLength={2}
      onChange={[Function]}
      valid={false}
      value=""
      withRef={true}
    />
    <Styled(styled.input)
      error={false}
      innerRef={[Function]}
      maxLength={4}
      onChange={[Function]}
      valid={false}
      value=""
      withRef={true}
    />
  </styled.div>
  <styled.div
    show={false}
  >
    Not a valid SSN
  </styled.div>
</styled.div>
`;

exports[`<Fields /> should render without blowing up 2`] = `
<styled.div>
  <Styled(styled.div)
    error="Error"
    valid={false}
  >
    SSN
  </Styled(styled.div)>
  <styled.div>
    <Styled(styled.input)
      error="Error"
      maxLength={3}
      onChange={[Function]}
      valid={false}
      value=""
    />
    <Styled(styled.input)
      error="Error"
      innerRef={[Function]}
      maxLength={2}
      onChange={[Function]}
      valid={false}
      value=""
      withRef={true}
    />
    <Styled(styled.input)
      error="Error"
      innerRef={[Function]}
      maxLength={4}
      onChange={[Function]}
      valid={false}
      value=""
      withRef={true}
    />
  </styled.div>
  <styled.div
    show="Error"
  >
    Not a valid SSN
  </styled.div>
</styled.div>
`;
